### ISSUE TITLE
Issues installing any Win9x version...

### ISSUE THREAD

**Issue by StScream** [joncampbell123/dosbox-x#2336]
Good day all! I was hoping someone could help me solve this. I installed Win98se following the instructions provided on  publisher website. Everything seemed to go great until at some point during one of the expected reboots it asked for the Win98 cd. Explorer had not opened yet and the only drive available to select was my c drive. I did make sure that my iso was mounted at startup before booting windows and it was.  I did copy the files from the ISO to my virtual drive but, although it recognized the CAB files it would not load the drivers from them. So I decided to start over from scratch. Now, no matter which iso I use (95, 98 and ME) it goes through the scan and says it's copying files to the drive but it ends after that with a "error loading progman.exe" (or something like that). 
Any ideas? My system is Tuf gaming x570 board, Ryzen3700 processor 32g ram and GT1030 video 2g memory.

**Comment by Sethur** [comment]
I can confirm this issue on Mac Os 11.2.2 Big Sur with Dosbox 0.83.11 SDL1

just tried installing both Windows 95 OSR 2.1 and OSR 2.5 according to [this guide](https://dosbox-x.com/wiki/Guide%3AInstalling-Windows-95) directly from the mounted installation CD (since first copying the files resulted in an ever blinking cursor after running setup.exe) and all attempts ended in: `Error loading PROGMAN.EXE`

Tried again using a Windows 95 boot image and at least setup seems to run properly now.

**Comment by Wengier** [comment]
@rderooy Are you able to reproduce this issue following the Win9x install guide?

**Comment by rderooy** [comment]
@Wengier The last time, I did an install for a test was probably 3 weeks or so ago. I will see if I can find a moment later today when I'm back home to do another test with the latest code.

**Comment by rderooy** [comment]
@Wengier it does seem that something is broken. My automated script to install Win95 builds no longer works with the latest nightly. 

The failure is different for RTM and OSR1 and later versions.

With RTM and OSR1 starting SETUP.EXE directly from the DOSBox-X prompt, shows garbage on the screen but no error in the log.
![dossetup_000](https://user-images.githubusercontent.com/17701098/110697573-d32ae200-81ec-11eb-9472-101350581fa4.png)

With later Win95 releases (OSR2+), instead the moment you run SETUP.EXE it seems to just hang with the following message written to the log:
```
LOG: INT 21h WRITE warning: DX=fffeh CX=7206h exceeds 64KB, truncating to 7206h
```
The DOSBox-X menus still work, and I can reboot using the menu.

**Comment by rderooy** [comment]
Win98SE SETUP.EXE instead just crashes with a "Packed file is corrupt error"

**Comment by rderooy** [comment]
I did a git bisect, and this is the offending commit that causes the "Packed file is corrupt" error for Win98SE SETUP.EXE
I did not test if this also fixes the Win95 issue.

[6e3e4845a35db15e96daaf03dc9291e064da2a4a] SVN r4330

**Comment by rderooy** [comment]
It does not solve the issues with Win95. I will have to go further back it seems. But not today...

**Comment by Wengier** [comment]
For the "Packed file is corrupt" error with Windows 98 SETUP (related to SVN r4330), perhaps @Allofich wants to take a look at this?

**Comment by Allofich** [comment]
I reviewed that commit and found a few spots where I hadn't quite matched up the SVN code to the existing DOSBox-X code properly, although I doubt it's the cause of the problem here.
At
https://github.com/Allofich/dosbox-x/tree/test
there is a commit fixing those spots. Can you give it a try, @rderooy?

**Comment by rderooy** [comment]
@Allofich I compiled your tree, and get the same error with Win98SE.

**Comment by rderooy** [comment]
I must have made a mistake with my Win95 testing yesterday. I did another git bisect just now, and also ended up with the same  offending commit: 6e3e4845a35db15e96daaf03dc9291e064da2a4a when testing Win95 (RTM and OSR2).

So this commit break both Win95 and Win98 in different ways.

**Comment by joncampbell123** [comment]
Could it be part of the code that changes how the FAT chain is extended if writing at the current EOF?

**Comment by rderooy** [comment]
@joncampbell123 yes restoring that piece of code fixes the issue for both Win95 and Win98.

**Comment by joncampbell123** [comment]
The latest code restores the original version of that part of the code. Installing Windows 95 appears to work fine with it.

**Comment by rderooy** [comment]
Basically, if your **installing** Windows 9x either use 0.83.10 or compile the latest code. For **running** Windows 9x you can use 0.83.11.

I'm wondering if this issue should warrant a 'hot-fix' release, since the next release is still almost 20 days out...

**Comment by joncampbell123** [comment]
How about a pre-release this Sunday or Saturday?

**Comment by rderooy** [comment]
That works for me...

**Comment by StScream** [comment]
Going back to previous version did work for me. Thanks!

**Comment by DSN9990** [comment]
Thanks. I experienced the same error when installing Win95 OSR2. @Wengier Can you give me a updated x64 VS2019 SDL2 build?

**Comment by Wengier** [comment]
@DSN9990 Sure. You can download the updated x64 VS2019 SDL2 build from:

* [dosbox-x-64-update.zip](http://cndos.fam.cx/dosware/dosbox-x-64-update.zip)

**Comment by rderooy** [comment]
I just compiled the latest code, and when I start Win98SE setup I get a "Message SU991016"
![krnl386_000](https://user-images.githubusercontent.com/17701098/111084604-747aa680-8513-11eb-9acb-be41ee6f20e2.png)

In the log:
```
LOG: 2994010001 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 2994094874 ERROR BIOS:Disk 0 not active
LOG: 2994106907 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 2994156547 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 2994198342 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 2994294964 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 3002162368 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 3005499023 ERROR IOCTL:DOS:IOCTL Call  D:6F Drive  2 unhandled (CH=08h)
LOG: 3019748392 ERROR DOSMISC:DOS:IOCTL Call  5 unhandled
LOG: 3024896943 ERROR IOCTL:DOS:IOCTL Call  D:6E Drive  2 unhandled (CH=08h)
```

I tried it with both ``cputype=pentium_mmx`` and ``cputype=pentium`` and get the same error. So this does not seem to be MMX related.

**Comment by rderooy** [comment]
Win95 RTM does install without problems.

**Comment by Wengier** [comment]
@rderooy Are you able to start Win98SE setup without issues with any recent (post-0.83.11) code?

**Comment by rderooy** [comment]
@Wengier it is not a change in the last few days that causes this breakage. I just did a checkout of fa00ee2a99994310fba3ac54d1b251cc207104dd and it is also broken.

0.83.10 (commit 8c91d7d0b646a3440e4c669629841f265ed594a8) does work for installing Win98SE or Win98 OSR2.

So somewhere in between things broke. The problem is that if it broke after commit https://github.com/joncampbell123/dosbox-x/commit/6e3e4845a35db15e96daaf03dc9291e064da2a4a it becomes difficult to track down as git bisect by itself will not be enough, I will need to patch each bisect point to see if the issue is still present.

**Comment by rderooy** [comment]
I just confirmed, I compiled commit 5341d4c3a585c7eff74d0840e329f50886034a70 which is the commit just before the commit that caused the Win98SE Packed file corrupt error, and it works. So it broke somewhere in between

5341d4c3a585c7eff74d0840e329f50886034a70 (good) and fa00ee2a99994310fba3ac54d1b251cc207104dd (bad)

**Comment by rderooy** [comment]
```
$ git bisect start
$ git bisect bad fa00ee2a99994310fba3ac54d1b251cc207104dd
$ git bisect good 5341d4c3a585c7eff74d0840e329f50886034a70
Bisecting: 106 revisions left to test after this (roughly 7 steps)
[95b10b57e32b437fbdf80553d806ec83577b0f92] fix issue #2257
```

**Comment by rderooy** [comment]
git bisect finished...
```
7898dabc820df8cb9b4d7c8c9c28f02d45edc6b4 is the first bad commit
commit 7898dabc820df8cb9b4d7c8c9c28f02d45edc6b4
Author: Wengier <wengierwu@yahoo.com>
Date:   Tue Feb 16 19:47:46 2021 -0500

    Fix creating temporary files

 CHANGELOG             |  2 ++
 src/dos/dos_files.cpp | 18 +++++++++++++++++-
 src/shell/shell.cpp   | 12 ++----------
 3 files changed, 21 insertions(+), 11 deletions(-)
```

**Comment by Wengier** [comment]
@rderooy Thanks for find it out. With #2376 I was able to overcome the "Message SU991016" issue when installing Windows 98.

**Comment by rderooy** [comment]
@Wengier Thanks. Win98SE now indeed gets passed the SU991016 error. But I have yet to do a full-install of Win98SE or indeed any Win9x version to ensure there are no further issues. And that will not happen from my side until at least this evening.

**Comment by rderooy** [comment]
With the latest code from github, 'I have successfully tested installing (running setup.exe directly from the dosbox-x shell):
- Win95 RTM
- Win95 OSR 1
- Win95 OSR 2
- Win95 OSR 2.1
- Win95 OSR 2.5
- Win98
- Win98SE

I also did a install of WinME (cd-rom boot method), and that also worked. Although the install of WinME threw many hundreds of these kind of errors during installation it did finish fine:
```
LOG: Limit check 20d600+4-1 = 20d603 > ffff
LOG: Segment limit violation
LOG: Limit check 1fa000+4-1 = 1fa003 > ffff
LOG: Segment limit violation
```

**Comment by Wengier** [comment]
@rderooy Great to hear that all Windows 9x (even ME) installations now work as expected. I notice that in the [Windows ME Wiki page](https://dosbox-x.com/wiki/Guide%3AInstalling-Windows-ME) it still says "There is no specific guide at this point for installing Windows ME in DOSBox-X, as Windows ME is not currently supported". You may want to change the sentence to mention that Windows ME is indeed supported, but specific problems may be expected.

Also, I think you can add to the [NE2000 Wiki page](https://dosbox-x.com/wiki/Guide%3ASetting-up-networking-in-DOSBox%E2%80%90X) about the Slirp backend support which was added recently.

**Comment by rderooy** [comment]
I tried running WinME SETUP.EXE directly from the dosbox-x shell (with the /NM flag to bypass the error that you need at least a 150MHz CPU) after first coping the installer data to the HDD, and it stopped fairly soon after when it started with the file copy

![krnl386_000](https://user-images.githubusercontent.com/17701098/111672928-bad54b80-881a-11eb-9c19-48c9c2198dc7.png)

So it seems the CD-ROM Boot install method works, but the xcopy method does not.

**Comment by Skagnar** [comment]
I still get

![LockedUp](https://user-images.githubusercontent.com/80923597/111695039-067ef980-8809-11eb-8ae1-4dceb8ed700a.png)

In win10, after using the update posted on this thread.. (dates in files 3.14)...

*sigh* 
??

Thank You!,
Ed (Skagnar)

**Comment by rderooy** [comment]
You need to compile the latest code. The code from the 14th does not work if you look at when the issue was fixed.

**Comment by joncampbell123** [comment]
> I tried running WinME SETUP.EXE directly from the dosbox-x shell (with the /NM flag to bypass the error that you need at least a 150MHz CPU) after first coping the installer data to the HDD, and it stopped fairly soon after when it started with the file copy
> 
> ![krnl386_000](https://user-images.githubusercontent.com/17701098/111672928-bad54b80-881a-11eb-9c19-48c9c2198dc7.png)
> 
> So it seems the CD-ROM Boot install method works, but the xcopy method does not.

I get that exact same error as well with Windows ME.

**Comment by Skagnar** [comment]
> 
> 
> You need to compile the latest code. The code from the 14th does not work if you look at when the issue was fixed.

Woah woah... slow down here.. 

I'm not all about the compiling stuff myself! (I fall short of such knowledge lol)

Guess I'm waiting for the dosbox-x next release, eh? :-!

Thanks Again,
Ed.

**Comment by joncampbell123** [comment]
Following up on the crash with the debugger, I'm seeing the setup program call down to DOS by switching to real mode (not virtual 8086 mode as you'd expect), which means that this stage has a very primitive Win16 environment going there. In fact I don't think I ever saw it running in 32-bit protected mode. Are they using a cut down 286 "standard mode"?

I noticed when that crash happens, the last INT 21h call made is AX=71A0h, the LFN "get volume information" call which takes a drive specification and returns the filesystem ("FAT" in this case) into another buffer. I noticed that at the INT 21h call from protected mode, the pointers are set up properly, however when the call actually happens in real mode, the segment and offset point somewhere else and the INT 21h call sees gibberish for the path and "FAT" is written in some unrelated memory. So this crash may very well be some translation error by the protected mode environment causing memory corruption.

For some reason if I use IMGMAKE to make a FAT16 partition of about 1.6GB, and install to that, the crash does not happen. It only happens if IMGMAKE is used to make a FAT32 partition, or I install to a ~800MB disk image I already made sometime last year.

**Comment by joncampbell123** [comment]
if you run with xms=false, it runs it's own XMS manager (XMSMGR.EXE) but then after the prompt on what directory to install to, DOSBox-X exits with "E_Exit: No byte handler for write to fffffffe"

**Comment by joncampbell123** [comment]
Another oddity: If I boot my Windows 98 image to command prompt only, run the Win ME installer, I get segment limit violation messages that suggest all segments are 64KB long. Maybe there's something with DOSBox-X HIMEM.SYS emulation that is NOT setting segment limits to 64KB?

**Comment by joncampbell123** [comment]
Hm, you know why Windows ME's setup doesn't crash when booted from Windows 98?

In real mode Windows 98 DOS mode, INT 21h AX=71A0h returns an error (CF=1). Because it's real mode, and LFNs are not enabled in that mode.

**Comment by joncampbell123** [comment]
I can confirm disabling Long Filename emulation avoids that crash when running Windows ME's installer from the DOSBox-X DOS environment.

**Comment by Wengier** [comment]
@joncampbell123 Thanks for checking this out! I can also confirm that the same crash happens when booted from Windows 98 and I load DOSLFN before running the Windows ME Setup. So basically Windows ME DOS Setup does not expect the presence of an LFN API (either DOSLFN or another LFN driver), so disabling LFN emulation should allow Windows ME Setup to run properly. @rderooy You can certainly add this information to the Wiki.

**Comment by rderooy** [comment]
So your saying to add

```
[dos]
lfn=false
```
To the WinME instructions?

**Comment by Wengier** [comment]
@rderooy Yes. It is apparently a bug of the Windows ME Setup program (why it calls INT 21h AX=71A0h at all?), but setting "lfn=false" will hopefully workaround this bug.

**Comment by DSN9990** [comment]
@Wengier Since Win95 RTM up to Win98SE seems to work fine now as @rderooy said, could you make a new build as x64 VS2019 SDL2?

**Comment by rderooy** [comment]
I updated the draft WinME wiki guide.

Some issues that I ran into while testing
- I could not get NE2000 with ``backend=pcap`` going. WInME kept complaining that the adapter was not working. But ``backend=slirp`` did work. This needs more testing, but seems like this may be a regression.
- ``core=dynamic_x86`` causes a core dump on startup of WinME

**Comment by joncampbell123** [comment]
I also noticed that Windows ME's IDE driver is doing the INT 13h vm86 test in a way DOSBox-X cannot yet satisfy, which is why drive C: is always in "compatibility mode".

It's attempting reads 256 sectors long (count=0x00) which DOSBox-X attempts to emulate, but after the second sector, Windows ME aborts the read for some reason and moves on using INT 13h.

DOSBox-X emulates the ATA behavior where the C/H/S IDE registers directly reflect where it's reading from (which Windows 3.11 WfW 32-bit disk access requires), I wonder if Windows ME considers that proof the IDE controller is "too old" to work with.

**Comment by rderooy** [comment]
@joncampbell123 Yes you instructed in the past to always set ``int13fakev86io=false`` for WinME, which is reflected in the guide. Would be good to get to the bottom of the issue though.

**Comment by DSN9990** [comment]
@Wengier Could you give me a updated VS2019 x64 SDL2 build, please?

**Comment by Wengier** [comment]
@DSN9990 Sure. Try this:

* [dosbox-x-64-update.zip](http://cndos.fam.cx/dosware/dosbox-x-64-update.zip)

**Comment by Wengier** [comment]
@rderooy I just want to add that you can also select the CPU type as "Pentium Pro" to bypass the error that you need at least a 150MHz CPU when installing Windows ME, in addition to the /NM method.

**Comment by DSN9990** [comment]
@Wengier Can you make a new build as MinGW x64 SDL2? I tried a game on Win95 and it did not work, gives an error saying that failed to load a 3D model (related to the full 80-bit precision that VS2019 does not recognize).

**Comment by DSN9990** [comment]
@Wengier Like I said above. Can you give me a MinGW x64 SDL2 build, please?

**Comment by Wengier** [comment]
@DSN9990 Can you please try the updated MinGW x64 SDL1 build for now:

* [dosbox-x-mingw64.zip](http://cndos.fam.cx/dosware/dosbox-x-update.zip)

**Comment by PatTheMav** [comment]
> With the latest code from github, 'I have successfully tested installing (running setup.exe directly from the dosbox-x shell):
> 
> * Win95 RTM
> * Win95 OSR 1
> * Win95 OSR 2
> * Win95 OSR 2.1
> * Win95 OSR 2.5
> * Win98
> * Win98SE
> 
> I also did a install of WinME (cd-rom boot method), and that also worked. Although the install of WinME threw many hundreds of these kind of errors during installation it did finish fine:
> 
> ```
> LOG: Limit check 20d600+4-1 = 20d603 > ffff
> LOG: Segment limit violation
> LOG: Limit check 1fa000+4-1 = 1fa003 > ffff
> LOG: Segment limit violation
> ```

I'm currently installing Windows 98 SE following the instructions on the wiki and I get these segment limitation violations for the entire duration of the setup process - is this an unfortunate but "expected" error right now?

(Windows 98 SE Boot CD, macOS 12, dosbox-x build 20211130202141).

**Comment by grstatdoc-gr** [comment]
I'm not sure if this gets a new thread but, I'm understanding that I need to make a image file with IMGMAKE before installing Win98,
But, I get the IMGMAKE not found error:

<img width="860" alt="image" src="https://github.com/joncampbell123/dosbox-x/assets/71800985/6b76a9e5-2a65-48cd-a295-71826ae619c6">
 
So, I'm not sure what to try next

**Comment by grstatdoc-gr** [comment]
Some progress:
needed to use the command line in  DOSBoxpX rather than outside of it...
<img width="361" alt="image" src="https://github.com/joncampbell123/dosbox-x/assets/71800985/6692ba5e-c2c2-4c31-a09d-56928a440d4a">

**Comment by rderooy** [comment]
> I'm not sure if this gets a new thread but, I'm understanding that I need to make a image file with IMGMAKE before installing Win98, But, I get the IMGMAKE not found error:
> 
> <img alt="image" width="860" src="https://user-images.githubusercontent.com/71800985/244777843-6b76a9e5-2a65-48cd-a295-71826ae619c6.png">
> So, I'm not sure what to try next

Please don't hijack other people's issues. In any case, `imgmake` is a DOSBox-X command. From your screenshot it is clear that you are not on the DOSBox-X command-prompt, but that you have booted Win9x, and as such DOSBox-X commands are not available.

**Comment by grstatdoc-gr** [comment]
Fair enough...last time I started a new thread, I got a message that the issue already exists.
Personally, it's related to the existing one.
Would you make it a new issue, then?

**Comment by maron2000** [comment]
Anyway running a DOSBox-X command on a Windows command prompt will definitely not work LOL.

**Comment by grstatdoc-gr** [comment]
Yes, it didn't (I misunderstood that step about which command prompt to use)

Mike Dean
